Move static checks of layers[0] (e.g., isinstance check) to model build time

mikekgfb opened the pull request:
Summary: Move static checks of layers[0] (e.g., isinstance check) to model build time because isinstance() does not work for torchscripted code.  Because the validation is now performed while constructing the object, the isinstance() call is performed in eager mode at model build time, and we avoid needing to call  isinstance() at runtime to determine whether the layers in a model are an instance of the TransformerEncoderLayer class, or its derived classes.

Test Plan: sandcastle, github

Differential Revision: D46096222

mikaylagawarecki reviewed (commented):
Out of curiosity is the value set to `why_not_sparsity_fast_path` ever used to warn somewhere, or do we only use it in the conditional statements like [here](https://github.com/pytorch/pytorch/pull/102045/files#diff-2a3e6e69b9bee987dc595081ace31d17318fd808cedf4b1195dec72ed26503e9L273)?

mikekgfb commented:
> Out of curiosity is the value set to `why_not_sparsity_fast_path` ever used to warn somewhere, or do we only use it in the conditional statements like [here](https://github.com/pytorch/pytorch/pull/102045/files#diff-2a3e6e69b9bee987dc595081ace31d17318fd808cedf4b1195dec72ed26503e9L273)?

Only for debug by us or for users.  We could do it as 1 monolithic if (actually not quite - torchscript doesn't seem to honor short-cutting conditional python eval, so guarding checks should go into a separate if), but if the check fails, then the user has no clue why.

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -189,9 +190,32 @@ def __init__(self, encoder_layer, num_layers, norm=None, enable_nested_tensor=Tr
         self.num_layers = num_layers
         self.norm = norm
-        self.enable_nested_tensor = enable_nested_tensor
+        self.use_nested_tensor = enable_nested_tensor
```
This change would have BC implications, would be best not to change the name unless absolutely necessary

mikekgfb replied:
Yes... there's a tradeoff.  It has a changed semantic meaning, so I figure giving it the same name as the init argument might confuse the meaning.  (We sorta have that after I aligned the is_causal name between nn.Transformer* and nn.MHA with SDPA's name is_causal, but it's a hint for nn.MHA/nn.Transformer*)

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -238,24 +262,14 @@ def forward(
-            why_not_sparsity_fast_path = f"{str_first_layer} was not TransformerEncoderLayer"
-        elif first_layer.norm_first :
-            why_not_sparsity_fast_path = f"{str_first_layer}.norm_first was True"
+        if not hasattr(self, "use_nested_tensor"):
```
hmmm will this ever be the case? since we always set this in `__init__` and classes that inherit from this will probably call `super.__init__()`

mikekgfb replied:
Unfortunately that's not the case!  When we load a model in eager mode that was saved, we get the code that is loaded with import, but the objects that were saved.

mikaylagawarecki replied:
Gotcha

mikekgfb commented on torch/nn/modules/transformer.py:
```diff
@@ -406,6 +418,19 @@ class TransformerEncoderLayer(Module):
+    state.)
+
+    If your are implementing a custonm layer, you may derive it either from
+    the Module or TransformerEncoderLayer class.  If your custom layer can
+    take advantage of both torch.Tensors and Nested Tensors, make its
+    implementation a derived class of TransformerEncoderLayer. If your custom
+    Layer supports only torch.Tensor inputs, derive its implementation from
+    Module.
```
because of how isinstance works -- isinstance(instance_of_x, x)  returns True if instance_of_x is an instance of x, which includes any subclass.

mikekgfb commented:
@pytorchbbot merge

mikekgfb commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge -r

Commit:
Move isinstance check to model build time - and a few other checks too (#102045)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102045

Move isinstance check to model build time - and a few other checks too

Test Plan: sandcastle, github

Reviewed By: mikaylagawarecki, davidberard98

Differential Revision: D46096222

fbshipit-source-id: a1c306b1a7ac91b90b2fd446086c3cf6fd9922cb

mikekgfb commented:
@pytorchbot merge
